concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Updates.hs

```diff
@@ -605,6 +634,65 @@ processBakerStakeThresholdUpdates t bu = do
                     pendingUpdates = pendingUpdates{pBakerStakeThresholdQueue = newpQ}
                 }
 
+-- |Process the add anonymity revoker update queue.
+--  Ignores updates with duplicate ARs.
+processAddAnonymityRevokerUpdates :: (MonadBlobStore m)
+  => Timestamp
+  -> BufferedRef Updates
+  -> HashedBufferedRef ARS.AnonymityRevokers
+  -> m (Map.Map TransactionTime UpdateValue, BufferedRef Updates, HashedBufferedRef ARS.AnonymityRevokers)
+processAddAnonymityRevokerUpdates t bu hbar = do
+    u@Updates{..} <- refLoad bu
+    oldARs <- ARS.arRevokers <$> refLoad hbar
+    oldQ <- refLoad (pAddAnonymityRevokerQueue pendingUpdates)
+    let (ql, qr) = Seq.spanl ((<= t) . transactionTimeToTimestamp . fst) (uqQueue oldQ)
+    (changes, updatedARs) <- addAndAccumNonduplicateUpdates oldARs ARS.arIdentity UVAddAnonymityRevoker ql
+
+    newQ <- refMake oldQ { uqQueue = qr }
+    newU <- refMake u { pendingUpdates = pendingUpdates {pAddAnonymityRevokerQueue = newQ} }
+    updatedARsRef <- refMake . ARS.AnonymityRevokers $ updatedARs
+    return (changes, newU, updatedARsRef)
+
+-- |Process the add identity provider update queue.
+--  Ignores updates with duplicate IPs.
+processAddIdentityProviderUpdates :: (MonadBlobStore m)
+  => Timestamp
+  -> BufferedRef Updates
+  -> HashedBufferedRef IPS.IdentityProviders
+  -> m (Map.Map TransactionTime UpdateValue, BufferedRef Updates, HashedBufferedRef IPS.IdentityProviders)
+processAddIdentityProviderUpdates t bu hbip = do
+    u@Updates{..} <- refLoad bu
+    oldIPs <- IPS.idProviders <$> refLoad hbip
+    oldQ <- refLoad (pAddIdentityProviderQueue pendingUpdates)
+    let (ql, qr) = Seq.spanl ((<= t) . transactionTimeToTimestamp . fst) (uqQueue oldQ)
+    (changes, updatedIPs) <- addAndAccumNonduplicateUpdates oldIPs IPS.ipIdentity UVAddIdentityProvider ql
+
+    newQ <- refMake oldQ { uqQueue = qr }
+    newU <- refMake u { pendingUpdates = pendingUpdates {pAddIdentityProviderQueue = newQ} }
+    updatedIPsRef <- refMake . IPS.IdentityProviders $ updatedIPs
+    return (changes, newU, updatedIPsRef)
```

Comment:
If no changes are made, then we should not make new references, but return the old ones. This will be extremely wasteful as we will end up storing a new copy of the identity providers map for every (finalized) block. (This also applies to the anonymity revokers.)

Reply:
Ah, great catch! Should be fixed now 😊